cpp/src/cupti.cpp

```diff
@@ -95,20 +95,25 @@ void CuptiMonitor::start_monitoring() {
 }
 
 void CuptiMonitor::stop_monitoring() {
-    std::lock_guard<std::mutex> lock(mutex_);
+    {
+        std::lock_guard<std::mutex> lock(mutex_);
 
-    if (!monitoring_active_.load()) {
-        return;
-    }
+        if (!monitoring_active_.load()) {
+            return;
+        }
 
-    monitoring_active_.store(false);
+        monitoring_active_.store(false);
+    }
```

Comment:
I don't think you need this lock at all since reads of `monitoring_active_` are not otherwise protected by the mutex.

So:

```
stop_monitoring() {
  if (!is_monitoring()) {
      return;
  }
  monitoring_active_.store(false);
  if (sampling_thread_.joinable()) {
      sampling_thread_.join();
  }
  {
    std::lock_guard ...;
    capture_memory_usage_impl();
  }
  unsubscribe();
}
```
?